contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,516 @@
+# Scheduling Policy
+
+_Status: Draft_
+
+_Authors: @arnaudmz, @yastij_
+
+_Reviewers: @bsalamat, @liggitt_
+
+# Objectives
+
+-  Define the concept of scheduling policies
+-  Propose their initial design and scope
+
+## Non-Goals
+
+-  How taints / tolerations work.
+-  How NodeSelector works.
+-  How node / pod affinity / anti-affinity rules work.
+-  How several schedulers can be used within a single cluster.
+-  How priority classes work.
+-  How to set defaults in Kubernetes.
+
+# Background
+
+During real-life Kubernetes architecturing we encountered contexts where role-isolation (between administration and simple namespace usage in a multi-tenant context) could be improved. So far, no restriction is possible on toleration, priority class usage, nodeSelector, anti-affinity depending on user permissions (RBAC).
+
+Identified use-cases aim to ensure that administrators have a way to restrict users or namespaces, it allows administrators to:
+
+-  Restrict execution for specific applications (which are namespace scoped) into certain nodes
+-  Create policies that prevent users from even attempting to schedule workloads onto masters to maximize security
+-  require that a pods under a namespace run on dedicated nodes
+-  Restrict usage of some `PriorityClass`
+-  Restrict usage to a specific set of schedulers.
+-  placing pods on specific nodes (master roles for instance).
+-  using specific priority classes.
+-  expressing pod affinity or anti-affinity rules.
+
+Also Mandatory (and optionally default) values must also be enforced by scheduling policies in case of:
+
+-  mutli-arch (amd64, arm64) of multi-os (Linux, Windows) clusters
+-  multi-az / region / failure domain clusters
+
+# Overview
+
+### syntaxic standards
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: <POLICY_KIND>
+metadata:
+  name: <POLICY_NAME>
+spec:
+  bindingMode: <BINDING_MODE>            # Describes a bindingMode (any or all)
+  namespaces:                            # a list of namespaces that the policy targets (optional)
```

Comment:
Why not use `metav1.LabelSelector` ? Kubernetes' API favors label selector for API queries

Reply:
+1

As I know, currently it's very difficult (impossible) to select namespaces using names if no specific labels are created. For example, how to select the namespace `kube-system` without using labels? So I guess that's the intention behind that. But regarding the scenario using scheduling policy, I feel labelSelector should be sufficient for most of the use cases, based on explicit requirement for labeling, before the select_beyond_labels problem is solved.

Reply:
Restricting which pods get scheduled on master nodes to those in, most often `kube-system`, is one of the use cases we'd like to address with this policy type. I would prefer to start with the explicit list of namespaces over label selectors, if we only tackle one.

Reply:
@easeway exactly - the k8s API supports label selector query natively:

```
GET https://api.aws.seed.dev.k8s.ondemand.com/api/v1/namespaces?labelSelector=foo=bar
```

and there is no way to get multiple resources without doing multiple `GET` requests

Reply:
@ericavonb can you specify both `namespaces` and `namespaceSelector`?
If yes, then you do you do inner join or outer join of found namespaces?

Reply:
I was thinking it would be an outer join (union of namespaces in `namespaces` and who match a `namespaceSelector`). That is what makes most sense to me since if you have a finite list of namespaces, you can manually filter it down rather than requiring an additional label selector. @yastij did you have an intention here?

Reply:
I would think that we may restrict could not use both `namespaces` and `namespaceSelector` here, this brings much complexity.